python/samples.cpp

```diff
@@ -0,0 +1,155 @@
+#include <mutex>
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include <arbor/common_types.hpp>
+#include <arbor/sampling.hpp>
+#include <arbor/simulation.hpp>
+
+#include "error.hpp"
+#include "strprintf.hpp"
+
+namespace pyarb {
+
+// ToDo: trace entry of different types/container (e.g. vector of doubles to get all samples of a cell)
+
+struct trace_entry {
+    arb::time_type t;
+    double v;
+};
+
+// A helper struct (state) ensuring that only one thread can write to the buffer holding the trace entries (mapped by probe id)
+struct sampler_state {
+    std::mutex mutex;
+    std::unordered_map<arb::cell_member_type, std::vector<trace_entry>> buffer;
+
+    std::vector<trace_entry>& locked_sampler_vec(arb::cell_member_type pid) {
+        // lock the mutex, s.t. other threads cannot write
+        std::lock_guard<std::mutex> lock(mutex);
+        // return or create entry
+        return buffer[pid];
+    }
+
+    // helper function to search probe id in buffer
+    bool has_pid(arb::cell_member_type pid) {
+        std::unordered_map<arb::cell_member_type, std::vector<trace_entry>>::iterator it = buffer.find(pid);
+        return (it != buffer.end());
+    }
+
+    // helper function to push back to locked vector
+    void push_back(arb::cell_member_type pid, trace_entry value) {
+        auto& v = locked_sampler_vec(pid);
+        v.push_back(std::move(value));
+    }
+
+    // helper function to return whole buffer
+    const std::unordered_map<arb::cell_member_type, std::vector<trace_entry>>& samples() const {
+        return buffer;
+    }
+
+    // helper function to return trace entry of probe id
```

Comment:
This performs the same role as `probe_buffer`, but in a non-thread-safe manner.

Reply:
Done.